drivers: sensor: ntc-thermistor fixes

panikiel opened the pull request:
This patchset addresses multiple problems with the ntc-thermistor driver.

asemjonovs commented on drivers/sensor/ntc_thermistor/ntc_thermistor_calc.c:
```diff
@@ -79,18 +38,27 @@ int ntc_compensation_compare_ohm(const struct ntc_type *type, const void *key, c
  */
 static void ntc_lookup_comp(const struct ntc_type *type, unsigned int ohm, int *i_low, int *i_high)
```
This looks like my original implementation :-)  

The Zephyr community advised me to use the bsearch for space optimization.  The compare function pointer here should find the lower key where the value falls between.  I don't think this change is needed. See
https://github.com/zephyrproject-rtos/zephyr/pull/54947#pullrequestreview-1335999939 and https://pastebin.com/prpRVmjn

semihalf-anikiel-pawel replied:
Oh, I wasn't aware that the bsearch() approach was requested during review. Thanks for the link.

However, I don't understand how this is space optimization. The comparison function still takes up code.
I compared the generated code with and without this change (ARM build):
```
without this change:
  ntc_compensation_compare_ohm(): 62 bytes
  ntc_get_temp_mc(): 114 bytes
  compare_ohm_DT_DRV_COMPATx(): 16 bytes (per instance)
with this change:
  ntc_get_temp_mc(): 132 bytes
```
Am I missing something?

asemjonovs replied:
@cfriedt - looks like with implementing our own binary search algorithm we're saving space over using `bsearch`.  Any concerns with removing this?

cfriedt replied:
There are a lot of changes happening in this PR. It would be ideal to make an apples to apples comparison.

One aspect of using bsearch is that the search itself is less intermixed with other code that tweaks this or that. So it's maybe a bit more obvious what is happening. So maybe code is a bit more maintainable / less bug prone?

Also, if 5 different drivers used bsearch instead of implementing their own bsearch, then there is commonality savings.

It is not very difficult to reduce code size for one custom use case without bsearch of course.

E.g. for small arrays, the code size could be tiny and a linear search would be faster

semihalf-anikiel-pawel replied:
> One aspect of using bsearch is that the search itself is less intermixed with other code that tweaks this or that. So it's maybe a bit more obvious what is happening and code is maybe a bit more maintainable / less bug prone?

IMHO, this comparison function is much more complicated and error prone than a 7-line binary search implementation. 

> Also, if 5 different drivers used bsearch instead of implementing their own bsearch, then there is commonality savings.

That would happen if they wanted the find-exact behavior (and in that case yes, I agree using bsearch() is a good idea). But, if they want the find-largest-not-greater-than behavior, they would all have to implement the (again, IMO) overly complicated comparison function that "tricks" bsearch() into the other behavior.

> It is not very difficult to reduce code size for one custom use case without bsearch of course.

My intent with this change isn't to optimize code size (the size comparison I made was to question the original explanation for this approach), but to make the code more understandable than the original approach. If you disagree, I can keep the old one.

> E.g. for small arrays, the code size could be tiny and a linear search would be faster

Actually, is there a reason for not doing that? Are the compensation tables ever going to be really big?

cfriedt replied:
I'm not a sensor maintainer, so will defer to them.

asemjonovs commented on drivers/sensor/ntc_thermistor/ntc_thermistor_calc.c:
```diff
@@ -97,17 +97,17 @@ static void ntc_lookup_comp(const struct ntc_type *type, unsigned int ohm, int *
+ * @sample_mv: Measured voltage in mV
  */
-uint32_t ntc_get_ohm_of_thermistor(const struct ntc_config *cfg, uint32_t max_adc, int16_t raw_adc)
+uint32_t ntc_get_ohm_of_thermistor(const struct ntc_config *cfg, int sample_mv)
```
This is looking more like [linux implementation](https://github.com/torvalds/linux/blob/master/drivers/hwmon/ntc_thermistor.c#L382).  Lets add the additional safety checks here as well.

semihalf-anikiel-pawel replied:
Done

semihalf-anikiel-pawel commented:
The Run tests failure seems to be unrelated to my changes? I didn't touch PM (nor does this driver use PM).

semihalf-anikiel-pawel commented:
> Would you be able to fix the script as well? I was just made aware that this didn't get updated with an earlier PR. https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/utils/ntc_thermistor_table.py

Sure. I changed it to just output the zephyr,compensation-table property, it that fine?

asemjonovs commented on scripts/utils/ntc_thermistor_table.py:
```diff
@@ -46,24 +46,15 @@ def main(
 ) -> None:
     temps_range = range(temp_init, temp_final + interval - 1, interval)
 
-    print(f"/* NTC Thermistor Table Generated with {os.path.basename(__file__)} */\n")
-    print(f"/ {{")
-    print(
-        f"\tthermistor_R25_{int(r25)}_B_{int(beta)}: thermistor-R25-{int(r25)}-B-{int(beta)} {{"
```
Lets keep this portion.  The idea is to create a DT node that one can copy/paste into their device tree file.  We can drop the `status` and `compatible` like you've already done.

semihalf-anikiel-pawel replied:
Done

Commit:
drivers: sensor: ntc-thermistor: Fix voltage & resistance calculation

The pullup resistance was assumed to be the maximum ADC value. This is
incorrect when the reference ADC voltage and the thermistor pullup
voltage differ. Use the pullup_uv property from DT instead.

The equations for the thermistor resistance are also wrong. The correct
equations are (see https://en.wikipedia.org/wiki/Voltage_divider):
R1 = R2 * (Vin / Vout - 1)
for the positive-connected resistor, and
R2 = R1 * (Vin / Vout - 1)^-1
for the negative-connected resistor. These were transformed so that
they can be computed using integer math.

Signed-off-by: Paweł Anikiel <pan@semihalf.com>

Commit:
drivers: sensor: ntc-thermistor: Fix wrong n_comp value

The value of n_comp is wrong for ntc-thermistor-generic. Introduce a
helper macro to compute n_comp correctly for both thermistor types.

Signed-off-by: Paweł Anikiel <pan@semihalf.com>

Commit:
drivers: sensor: ntc-thermistor: Simplify comp lookup

bsearch() tries to find the exact match for the key, otherwise it
returns NULL. However, here we want to find two elements our key lies
in-between. This requires some rather complicated logic that pretends
it found the exact match even if it didn't. Replace this with a simple
implementation of binary search.

Signed-off-by: Paweł Anikiel <pan@semihalf.com>

Commit:
scripts: Update ntc-thermistor table generator

Update the script to work with the current version of the
ntc-thermistor driver.

Signed-off-by: Paweł Anikiel <pan@semihalf.com>

Commit:
drivers: sensor: ntc-thermistor: Remove r25-ohm property

Remove the r25-ohm property. It is not used by the driver, and it is
not present in linux.

Signed-off-by: Paweł Anikiel <pan@semihalf.com>

Commit:
drivers: sensor: ntc-thermistor: Add Murata NCP15WB473 thermistor

Add murata,ncp15wb473 property together with a compensation table. The
table can be found on the vendor's website
(Resistance VS. Temperature tab):
https://www.murata.com/en-us/products/productdetail?partno=NCP15WB473F03RC

Signed-off-by: Paweł Anikiel <pan@semihalf.com>

Commit:
drivers: sensor: ntc-thermistor: Add PM support

Add support for runtime power management.

Signed-off-by: Paweł Anikiel <pan@semihalf.com>

semihalf-anikiel-pawel commented on drivers/sensor/ntc_thermistor/ntc_thermistor.c:
```diff
@@ -35,6 +37,11 @@ static int ntc_thermistor_sample_fetch(const struct device *dev, enum sensor_cha
 		.calibrate = false,
 	};
 
+	res = pm_device_runtime_get(dev);
+	if (res) {
+		return res;
+	}
+
```
If `CONFIG_PM_DEVICE_RUNTIME` isn't defined, this function becomes a stub that returns zero, which should be optimized out: https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/pm/device_runtime.h#L179. Same for the `_put` below.

semihalf-anikiel-pawel commented:
@asemjonovs Could you please take another look? I made some changes to the PM logic and that dismissed your approval.

fabiobaltieri commented:
@semihalf-anikiel-pawel this needs a rebase as well btw

semihalf-anikiel-pawel commented on drivers/sensor/ntc_thermistor/ntc_thermistor.c:
```diff
@@ -92,9 +100,33 @@ static int ntc_thermistor_init(const struct device *dev)
+
+	err = pm_device_runtime_enable(dev);
+	if (err) {
+		LOG_ERR("Failed to enable runtime power management");
```
Yes, that's definitely better. Done

MaureenHelm reviewed (changes requested):
Except for the DT_INST concern, the rest of the PR looks fine

semihalf-anikiel-pawel commented on drivers/sensor/ntc_thermistor/ntc_thermistor.c:
```diff
@@ -95,45 +95,42 @@ static int ntc_thermistor_init(const struct device *dev)
 	return 0;
 }
 
-#define NTC_THERMISTOR_DEFINE0(inst, id, _comp, _n_comp)                                           \
-	static struct ntc_thermistor_data ntc_thermistor_driver_##id##inst;                        \
+#define NTC_THERMISTOR_DEFINE0(node_id, _comp, _n_comp)                                            \
+	static struct ntc_thermistor_data ntc_thermistor_data_##node_id;                           \
                                                                                                    \
-	static const struct ntc_thermistor_config ntc_thermistor_cfg_##id##inst = {                \
-		.adc_channel = ADC_DT_SPEC_INST_GET(inst),                                         \
+	static const struct ntc_thermistor_config ntc_thermistor_config_##node_id = {              \
+		.adc_channel = ADC_DT_SPEC_GET(node_id),                                           \
 		.ntc_cfg =                                                                         \
 			{                                                                          \
-				.pullup_uv = DT_INST_PROP(inst, pullup_uv),                        \
-				.pullup_ohm = DT_INST_PROP(inst, pullup_ohm),                      \
-				.pulldown_ohm = DT_INST_PROP(inst, pulldown_ohm),                  \
-				.connected_positive = DT_INST_PROP(inst, connected_positive),      \
+				.pullup_uv = DT_PROP(node_id, pullup_uv),                          \
+				.pullup_ohm = DT_PROP(node_id, pullup_ohm),                        \
+				.pulldown_ohm = DT_PROP(node_id, pulldown_ohm),                    \
+				.connected_positive = DT_PROP(node_id, connected_positive),        \
```
I was looking at these two drivers that also declare multiple compatibles:
- `drivers/eeprom/eeprom_at2x.c`
- `drivers/display/ssd16xx.c`

I tried to be consistent with how they do it. What are your thoughts about this?

MaureenHelm replied:
There are examples of both styles in the tree and some disagreement about which way to do it. Until there's a project-wide decision on a preferred style, I'd like to stick with `DT_INST` for now.

semihalf-anikiel-pawel replied:
Okay, I brought back the `DT_INST` style declarations.

semihalf-anikiel-pawel commented on drivers/sensor/ntc_thermistor/ntc_thermistor.c:
```diff
@@ -39,7 +39,7 @@ static int ntc_thermistor_sample_fetch(const struct device *dev, enum sensor_cha
 	adc_sequence_init_dt(&cfg->adc_channel, &sequence);
 	res = adc_read(cfg->adc_channel.dev, &sequence);
-	if (res) {
+	if (!res) {
```
Should I put it into a seperate commit?

MaureenHelm replied:
Yes

semihalf-anikiel-pawel replied:
Done

Commit:
drivers: sensor: ntc-thermistor: Fix flipped condition

Fix the flipped condition inside this if statement.

Signed-off-by: Paweł Anikiel <pan@semihalf.com>
